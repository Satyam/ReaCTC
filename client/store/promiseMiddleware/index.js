import omit from 'lodash/omit';

export const REQUEST_SENT = 'Stage: request sent';
export const REPLY_RECEIVED = 'Stage: reply received';
export const FAILURE_RECEIVED = 'Stage: failure received';

export default function promiseMiddleware() {
  return next => (action) => {
    if (action.promise && typeof action.promise.then === 'function') {
      const promise = action.promise;
      const act = omit(action, 'promise');
      next({
        ...act,
        stage: REQUEST_SENT,
      });
      return promise.then(
        response =>
          next({
            ...act,
            stage: REPLY_RECEIVED,
            payload: {
              ...act.payload,
              ...response,
            },
          }),
        (error) => {
          next({
            ...act,
            stage: FAILURE_RECEIVED,
            error: error.toString(),
          });
        }
      );
    }
    return next(action);
  };
}
