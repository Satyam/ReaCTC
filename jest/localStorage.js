const storage = {};

export default {
  setItem: (key, value) => {
    storage[key] = value || '';
  },
  getItem: key => storage[key] || null,
  removeItem: (key) => {
    delete storage[key];
  },
  get length() {
    return Object.keys(storage).length;
  },
  key: (i) => {
    const keys = Object.keys(storage);
    return keys[i] || null;
  },
};
