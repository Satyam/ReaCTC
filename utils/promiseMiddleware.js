// @flow
import omit from 'lodash/omit';

export const REQUEST_SENT = 'Stage: request sent';
export const REPLY_RECEIVED = 'Stage: reply received';
export const FAILURE_RECEIVED = 'Stage: failure received';

type Action = {
  +type: string,
};

type PromiseAction = {
  +type: string,
  promise: ?Promise<any>,
};

type Next = (action: Action) => any;
export default function promiseMiddleware() {
  return (next: Next) => (action: PromiseAction) => {
    if (action.promise && typeof action.promise.then === 'function') {
      const promise = action.promise;
      const act = omit(action, 'promise');
      next({
        ...act,
        stage: REQUEST_SENT,
      });
      return promise.then(
        response =>
          next({
            ...act,
            stage: REPLY_RECEIVED,
            payload: Object.assign(
              {},
              act.payload,
              Array.isArray(response) ? { list: response } : response
            ),
          }),
        (error) => {
          next({
            ...act,
            stage: FAILURE_RECEIVED,
            error: error.toString(),
          });
        }
      );
    }
    return next(action);
  };
}
