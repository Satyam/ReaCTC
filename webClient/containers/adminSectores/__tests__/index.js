import React from 'react';
import { shallow, mount } from 'enzyme';
import jestFetchMock from 'jest-fetch-mock';
import configureStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';
import promiseMiddleware from '_utils/promiseMiddleware';
import localStorage from '_jest/localStorage';

import HoC, { mapStateToProps, mapDispatchToProps, storeInitializer } from '..';

global.fetch = jestFetchMock;
global.localStorage = localStorage;

const listSectores = [
  {
    idSector: 'constitucion',
    descrCorta: 'Constitución',
    descr: 'Estación Constitución, Ciudad de Buenos Aires, Argentina',
  },
  {
    idSector: 'retiro',
    descrCorta: 'Retiro',
    descr: 'Mentira, apenas una línea',
  },
];

describe('mapStateToProps', () => {
  it('should be a function', () => {
    expect(mapStateToProps).toBeInstanceOf(Function);
  });
  it('with no data', () => {
    const state = {
      sectores: {
        list: {
          list: null,
        },
        adminStatus: null,
      },
    };
    const map = mapStateToProps(state);
    expect(map.sectores).toBeNull();
    expect(map.status).toBeNull();
  });
  it('with arbitrary data', () => {
    const state = {
      sectores: {
        list: {
          list: [1, 2],
        },
        adminStatus: [3, 4],
      },
    };
    const map = mapStateToProps(state);
    expect(map.sectores).toEqual([1, 2]);
    expect(map.status).toEqual([3, 4]);
  });
});
describe('mapDispatchToProps', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });
  it('should be a function', () => {
    expect(mapDispatchToProps).toBeInstanceOf(Function);
  });
  it('should return object with expected keys', () => {
    const dispatch = jest.fn();
    const map = mapDispatchToProps(dispatch);
    const keys = Object.keys(map);
    expect(keys).toContain('onDeleteSectores');
    expect(keys).toContain('onUploadSector');
    expect(keys).toContain('onClearStatusAdmin');
    expect(map.onDeleteSectores).toBeInstanceOf(Function);
    expect(map.onUploadSector).toBeInstanceOf(Function);
    expect(map.onClearStatusAdmin).toBeInstanceOf(Function);
  });
  it('mapDispatchToProps - onDeleteSectores', () => {
    const fetchMock = fetch.mockResponse(JSON.stringify({}), { status: 200 });
    const mockStore = configureStore([reduxThunk, promiseMiddleware]);
    const store = mockStore({ sectores: { list: { requested: [4, 5] } } });

    const map = mapDispatchToProps(store.dispatch);
    // This one is async
    return map.onDeleteSectores([1, 2, 3]).then((data) => {
      expect(data).toMatchSnapshot('mapDispatchToProps - onDeleteSectores - returned data');
      expect(fetchMock.mock.calls).toMatchSnapshot(
        'mapDispatchToProps - onDeleteSectores - fetchMock calls'
      );
      expect(store.getActions()).toMatchSnapshot(
        'mapDispatchToProps - onDeleteSectores - store getActions'
      );
    });
  });
  it('mapDispatchToProps - onClearStatusAdmin', () => {
    const fetchMock = fetch.mockResponse(JSON.stringify({}), { status: 200 });
    const mockStore = configureStore([reduxThunk, promiseMiddleware]);
    const store = mockStore({});

    const map = mapDispatchToProps(store.dispatch);
    // This one is synchronous
    const data = map.onClearStatusAdmin();
    expect(data).toMatchSnapshot('mapDispatchToProps - onClearStatusAdmin - returned data');
    expect(fetchMock.mock.calls).toMatchSnapshot(
      'mapDispatchToProps - onClearStatusAdmin - fetchMock calls'
    );
    expect(store.getActions()).toMatchSnapshot(
      'mapDispatchToProps - onClearStatusAdmin - store getActions'
    );
  });
});
describe('storeInitializer', () => {
  it('should be a function', () => {
    expect(storeInitializer).toBeInstanceOf(Function);
  });
  it('should return object with expected keys', () => {
    const fetchMock = fetch.mockResponse(JSON.stringify(listSectores), { status: 200 });
    const mockStore = configureStore([reduxThunk, promiseMiddleware]);
    const store = mockStore({ sectores: { list: { list: [], requested: false } } });

    return storeInitializer(store.dispatch).then((data) => {
      expect(data).toMatchSnapshot('storeInitializer - data');
      expect(fetchMock.mock.calls).toMatchSnapshot('storeInitializer - fetchMock calls');
      expect(store.getActions()).toMatchSnapshot('storeInitializer - store getActions');
    });
  });
});
describe('composition', () => {
  it('regular render', () => {
    const fetchMock = fetch.mockResponse(JSON.stringify(listSectores), { status: 200 });
    const mockStore = configureStore([reduxThunk, promiseMiddleware]);
    const store = mockStore({ sectores: { list: { list: [], requested: false } } });

    expect(shallow(<HoC />, { context: { store } })).toMatchSnapshot();
    expect(fetchMock.mock.calls).toMatchSnapshot('composition - fetchMock calls');
    expect(store.getActions()).toMatchSnapshot('composition- store getActions');
  });
});
