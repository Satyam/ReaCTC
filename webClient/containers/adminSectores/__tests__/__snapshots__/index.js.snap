// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition - fetchMock calls 1`] = `
Array [
  Array [
    "http://localhost:8080/data/v2/sectores/",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Accept": "application/json",
        "Authorization": null,
        "Content-Type": "application/json",
      },
      "method": "get",
    },
  ],
  Array [
    "http://localhost:8080/data/v2/sectores/",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Accept": "application/json",
        "Authorization": null,
        "Content-Type": "application/json",
      },
      "method": "get",
    },
  ],
]
`;

exports[`composition- store getActions 1`] = `
Array [
  Object {
    "stage": "Stage: request sent",
    "type": "sectores / list sectores",
    "wsMode": false,
  },
]
`;

exports[`mapDispatchToProps - onClearStatusAdmin - fetchMock calls 1`] = `Array []`;

exports[`mapDispatchToProps - onClearStatusAdmin - returned data 1`] = `
Object {
  "type": "sectores / clear status admin",
}
`;

exports[`mapDispatchToProps - onClearStatusAdmin - store getActions 1`] = `
Array [
  Object {
    "type": "sectores / clear status admin",
  },
]
`;

exports[`mapDispatchToProps - onDeleteSectores - fetchMock calls 1`] = `
Array [
  Array [
    "http://localhost:8080/data/v2/sectores/1,2,3",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Accept": "application/json",
        "Authorization": null,
        "Content-Type": "application/json",
      },
      "method": "delete",
    },
  ],
]
`;

exports[`mapDispatchToProps - onDeleteSectores - returned data 1`] = `undefined`;

exports[`mapDispatchToProps - onDeleteSectores - store getActions 1`] = `
Array [
  Object {
    "payload": Object {
      "idSectores": Array [
        1,
        2,
        3,
      ],
    },
    "stage": "Stage: request sent",
    "type": "sectores / delete sector",
  },
  Object {
    "payload": Object {
      "idSectores": Array [
        1,
        2,
        3,
      ],
    },
    "stage": "Stage: reply received",
    "type": "sectores / delete sector",
  },
  Object {
    "payload": Object {
      "entity": "1,2,3",
      "message": "Borrados",
      "nivel": "normal",
    },
    "type": "sectores / add status admin",
  },
]
`;

exports[`storeInitializer - data 1`] = `
Object {
  "payload": Object {
    "list": Array [
      Object {
        "descr": "Estación Constitución, Ciudad de Buenos Aires, Argentina",
        "descrCorta": "Constitución",
        "idSector": "constitucion",
      },
      Object {
        "descr": "Mentira, apenas una línea",
        "descrCorta": "Retiro",
        "idSector": "retiro",
      },
    ],
  },
  "stage": "Stage: reply received",
  "type": "sectores / list sectores",
  "wsMode": false,
}
`;

exports[`storeInitializer - fetchMock calls 1`] = `
Array [
  Array [
    "http://localhost:8080/data/v2/sectores/",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Accept": "application/json",
        "Authorization": null,
        "Content-Type": "application/json",
      },
      "method": "get",
    },
  ],
]
`;

exports[`storeInitializer - store getActions 1`] = `
Array [
  Object {
    "stage": "Stage: request sent",
    "type": "sectores / list sectores",
    "wsMode": false,
  },
  Object {
    "payload": Object {
      "list": Array [
        Object {
          "descr": "Estación Constitución, Ciudad de Buenos Aires, Argentina",
          "descrCorta": "Constitución",
          "idSector": "constitucion",
        },
        Object {
          "descr": "Mentira, apenas una línea",
          "descrCorta": "Retiro",
          "idSector": "retiro",
        },
      ],
    },
    "stage": "Stage: reply received",
    "type": "sectores / list sectores",
    "wsMode": false,
  },
]
`;
